### DISCUSSION TITLE
Interface for using external diffing tools

### DISCUSSION THREAD

**Discussion by seanfarley** [magit/magit#4777] [Ideas]
Tassilo Horn wrote a [fantastic blog post](https://www.tsdh.org/posts/2022-08-01-difftastic-diffing-with-magit.html) about getting [difftastic](https://github.com/Wilfred/difftastic) output into magit. After some private discussion with them and others, I was wondering if we could open a discussion about formalizing an interface with magit so that external diff tools could be used as plugins. From the discussion, some design points came up:

- displaying hunks within the familiar diff sections list (to allow jumping around files and hunks)
- not relying (as the blog post does) on internal functions
- probably lots of other things that I forgot

This discussion is just to open up dialog about how welcomed an interface like this would be. Or perhaps this is part of a larger system of having a plugin api in general.

**Comment by verhovsky** [comment]
difftastic parses files and diffs abstract syntax trees (ASTs), so it can generate nicer diffs than the default of just diffing raw lines/characters.

I wanted to use difftastic system-wide by putting this in my `~/.gitconfig`

```
[diff]
	external = difft
```

but this causes Magit's diff highlighting to stop working when looking at previous commits or stashes. Magit just displays difftastic's output as-is and doesn't show colors (which difftastic outputs). So it's unusable, I have to pick either Magit or difftastic.

**Reply by tarsius** [reply]
Use `magit-git-global-arguments` to disable settings in just Magit.

**Reply by verhovsky** [reply]
ChatGPT told me you're telling me to do this

```emacs
(after! magit
  (setq magit-git-global-arguments
    (append magit-git-global-arguments '("--no-ext-diff"))))
```

but I get a "Corrupt Git configuration" when I open Doom Emacs.

**Reply by tarsius** [reply]
Not every git command accepts the `--no-ext-diff` argument -- those that do not deal with diffs at all, have no reason to do so.

As the doc-string points out, you can only use arguments that can be used in between `git` and `<subcommand>` and the valid arguments are listed in the `git(1)` manpage. Looking at the default value, you will notice that the `-c` argument is used a lot. So what you have to do now, is to figure out what variable=value corresponds to the `--no-ext-diff` argument. The `git-diff(1)` and `git-config(1)` are good places to start, one of them is bound to provide that information.

**Comment by tarsius** [comment]
I hope that this summer or fall I will finally get around to looking at difftastic more closely.

**Reply by ypytomets** [reply]
https://github.com/pkryger/difftastic.el render output quite well.  The only thing missing is to use it as magit's default.
